### DISCUSSION TITLE
STM32H7 Zephyr CAN Driver

### DISCUSSION THREAD

**Discussion by RIPHeisenberg** [zephyrproject-rtos/zephyr#69073] [General] [answered]
Hello everyone, 

I currently started with the STM32H747I_disco and CAN. The CAN implementation looks like this:

fdcan1: can@4000a000 {
			compatible = "st,stm32h7-fdcan";
			reg = <0x4000a000 0x400>, <0x4000ac00 0x350>;
			reg-names = "m_can", "message_ram";
			clocks = <&rcc STM32_CLOCK_BUS_APB1_2 0x00000100>;
			interrupts = <19 0>, <21 0>, <63 0>;
			interrupt-names = "LINE_0", "LINE_1", "CALIB";
			bosch,mram-cfg = <0x0 28 8 3 3 0 3 3>;
			sample-point = <875>;
			sample-point-data = <875>;
			status = "disabled";
};

I added following devicetree overlay file:

/ {
  chosen {
    zephyr,console = &arduino_serial;
    zephyr,canbus = &fdcan1;
  };

  can_phy2: can-phy2 {
    compatible = "can-transceiver-gpio";
    enable-gpios = <&gpiob 8 GPIO_ACTIVE_LOW>;
    standby-gpios = <&gpiob 9 GPIO_ACTIVE_HIGH>;
    max-bitrate = <8000000>;
    #phy-cells = <0>;
  };  
};

&fdcan1 {
  compatible = "st,stm32h7-fdcan";
  bus-speed = <125000>;
	bus-speed-data = <1000000>;
	status = "okay";
  phys = <&can_phy2>;
};

`void CAN_init(void)
{
	can_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_canbus));
	if (!device_is_ready(can_dev)) {
		printk("CAN: Device driver not ready.\n");
		return 0;
	}
}`

The debug terminal sends a error message: 

"can_stm32h7: CAN pinctrl setup failed (-2)"

I want to use the onboard CAN drivers but there isn't a native pinctrl configuration available in the YAML file. Some users have used different drivers like SPI to communicate with a external CAN driver this is not what I want. Am I getting something fundamentally wrong? Is it possible to use STM32 native CAN drivers with Zephyr. I could switch back to the STM32 hal drivers, but I'd rather not.

Informations:
STM32H747i_DISCOVERY
Zephyr 3.5.99
Zephyr SDK 0.16.5
commit: 023d4838fbb8513f9d9ff6f26b065591131c347d

I'm compiling and running the project on the M4 chip.

I hope this information is sufficient to get ahead. Its my first post and I'd really like to get in touch with the community.

Best regards

**Comment by henrikbrixandersen** [comment]
> "can_stm32h7: CAN pinctrl setup failed (-2)"

Everything in your DTS snippets above look fine, except that you're missing a pinctrl configuration in `&fdcan1 { ... }` (and you shouldn't override `compatible` there). Try something like the following, replacing the pinctrl pins with the actual pins routed to your CAN transceiver:

```
&fdcan1 {
        pinctrl-0 = <&fdcan1_rx_pb8 &fdcan1_tx_ph13>;
        pinctrl-names = "default";
        bus-speed = <125000>;
        bus-speed-data = <1000000>;
        status = "okay";
        phys = <&can_phy2>;
};
```

Current `main` (and the upcoming Zephyr v3.6.0) furthermore supports specifying the domain clock for FDCAN, a feature you may or may not need depending on your board clock configuration. See commit bc69500b0e101087d5c9f7261c2e729882aee88a for further details.

> Am I getting something fundamentally wrong? Is it possible to use STM32 native CAN drivers with Zephyr. I could switch back to the STM32 hal drivers, but I'd rather not.

It is definitely possible, yes. I am using the FDCAN driver for the STM32H7 myself.

**Reply by RIPHeisenberg** [reply]
Hi Hendrik,

thanks for the swift responds. The pinctrl macros that you have defined look fairly nice. Those are however not defined in my HAL Files. I tried to add them by myself but it seems as if I'm doing something wrong. The compiler doesn't like this.

ERR point:
DT_INST_FOREACH_STATUS_OKAY(CAN_STM32H7_MCAN_INIT)
 
 My .overlay looks like this now: 
`
&pinctrl {
  fdcan1_rx_pb8: fdcan1_rx_pb8 {
    pinmux = <STM32_PINMUX('B', 8, AF9)>;
  };
  fdcan1_tx_pb9: fdcan1_tx_pb9 {
    pinmux = <STM32_PINMUX('B', 9, AF9)>;
  };
};


&fdcan1 {
  pinctrl-0 = <&fdcan1_rx_pb8 &fdcan1_tx_pb9>;
  bus-speed = <125000>;
  bus-speed-data = <1000000>;
  status = "okay";
};
`

PB8 and PB9 are the official initialized pins for FDCAN1. I have seen the fdcan pinctrl definitions in other STM32 boards but can't find them within my STM32H747I .dts files.

Is it possible to add them by myself are there defined somewhere else? What am I missing.

Best regards

**Reply by henrikbrixandersen** [reply]
They are defined in `../modules/hal/stm32/dts/st/h7/stm32h747xihx-pinctrl.dtsi`.

**Reply by RIPHeisenberg** [reply]
So this is probably a noob question but how can I add the stm32h747xihx-pinctrl.dtsi to my devicetree overlay? It is not a sub folder from my zephyr but in parallel to the west workspace like:

west_workspace:
-> modules/hal/...
-> zephyr/...


I copied the devicetree definitions to my overlay file to see if it might work. The overlay file now looks like this
`&pinctrl {
  fdcan1_rx_pb8: fdcan1_rx_pb8 {
    pinmux = <STM32_PINMUX('B', 8, AF9)>;
  };
  fdcan1_tx_pb9: fdcan1_tx_pb9 {
    pinmux = <STM32_PINMUX('B', 9, AF9)>;
  };
};
 


&fdcan1 {
  pinctrl-0 = <&fdcan1_rx_pb8 &fdcan1_tx_pb9>;
  bus-speed = <125000>;
  bus-speed-data = <1000000>;
  status = "okay";
};
`

It still doesn't compile. My kconfig looks like this:
CONFIG_CAN=y
CONFIG_CAN_FD_MODE=y
CONFIG_CAN_STM32H7_FDCAN=y

am I missing something? 

thanks a lot hendrik :D

**Reply by erwango** [reply]
Likely your hal_stm32 module is outdated.
File mentioned by @henrikbrixandersen is available in recent version of the repo, which you see by running a `west update` command.

**Reply by RIPHeisenberg** [reply]
Hey thanks for the hint. I added hal_stm32 to my west.yml file. I wasn't aware that this was mandatory. Now I'm getting the devicetree definitions. Very nice :D

the code however still doesn't compile...

My devicetree looks like this now:

fdcan1: can@4000a000 {
    compatible = "st,stm32h7-fdcan";
    reg = <0x4000a000 0x400>, <0x4000ac00 0x350>;
    reg-names = "m_can", "message_ram";
    clocks = <&rcc STM32_CLOCK_BUS_APB1_2 0x00000100>;
    interrupts = <19 0>, <21 0>, <63 0>;
    interrupt-names = "LINE_0", "LINE_1", "CALIB";
    bosch,mram-cfg = <0x0 28 8 3 3 0 3 3>;
    sample-point = <875>;
    sample-point-data = <875>;
    status = "okay";
    pinctrl-0 = <&fdcan1_rx_pb8 &fdcan1_tx_pb9>;
    bus-speed = <125000>;
    bus-speed-data = <1000000>;
};

Is there something wrong with it?
Err:
In: /west_workspace/zephyr/drivers/can/can_stm32h7_fdcan.c

error: 'PINCTRL_STATE_DT_N_S_soc_S_can_4000a000_PINCTRL_IDX_0_UPPER_TOKEN' undeclared here (not in a function); did you mean 'DT_N_S_soc_S_spi_40015000_PINCTRL_IDX_0_UPPER_TOKEN'?
   91 |         _CONCAT(PINCTRL_STATE_,                                                \
      |                 ^~~~~~~~~~~~~~


Do I have to add additional kconfig configurations?

**Reply by erwango** [reply]
At least, you're missing following property: pinctrl-names = "default";.
But I'm surprised this error is not caught earlier actually.

**Reply by RIPHeisenberg** [reply, marked as answer]
That actually was the issue. Thanks! It finally compiles now. And it seems as if packages are being send. :D I'll have to test it with a can module but I think this should work out. 

1. Adding stm32_hal to zephyr west.yml  file

Final overlay file:

 #include <st/h7/stm32h747Xi_m4.dtsi>

 / {
  chosen {
    zephyr,console = &uart8;
    zephyr,canbus = &fdcan1;
  };
 };

 &uart8 {
  status="okay";
 };
 
 &fdcan1 {
  pinctrl-names = "default";
  pinctrl-0 = <&fdcan1_rx_pb8 &fdcan1_tx_pb9>;
  bus-speed = <125000>;
  bus-speed-data = <1000000>;
  status = "okay";
};

Kconfig:

CONFIG_GPIO=y
CONFIG_CAN=y
CONFIG_CAN_FD_MODE=y
CONFIG_CAN_STM32H7_FDCAN=y

CONFIG_ISOTP=y
CONFIG_REBOOT=y

CONFIG_LOG=y
CONFIG_DEBUG=y
CONFIG_CONSOLE=y
CONFIG_PRINTK=y
CONFIG_UART_CONSOLE=y
CONFIG_LOG_BACKEND_UART=y

CONFIG_SAMPLE_RX_THREAD_STACK_SIZE=1024
CONFIG_SAMPLE_RX_THREAD_PRIORITY=7

# We have two receiving contexts that are bound to a single address.
CONFIG_ISOTP_RX_SF_FF_BUF_COUNT=2

It might help someone else in the future. Thanks for the support <3 

Cheers
